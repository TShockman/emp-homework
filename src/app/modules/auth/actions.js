export const NAME = 'auth';

// lifecycle action names for a login request
export const LOGIN_REQUESTED = `${NAME}/LOGIN_REQUESTED`;
export const LOGIN_FULFILLED = `${NAME}/LOGIN_FULFILLED`;
export const LOGIN_REJECTED = `${NAME}/LOGIN_REJECTED`;
// NOTE: currently the rejected action is unused for simplicity sake
// since logging in always succeeds anyways. In a fuller implementation,
// LOGIN_REJECTED would be dispatched if the service request failed, the
// auth reducer would update the auth state to signify this, and the Login
// component would handle displaying relevant error text to the user.