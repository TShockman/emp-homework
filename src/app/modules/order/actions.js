export const NAME = 'order';

// action names for get orders flow
export const GET_ORDERS_REQUESTED = `${NAME}/GET_ORDERS_REQUESTED`;
export const GET_ORDERS_FULFILLED = `${NAME}/GET_ORDERS_FULFILLED`;
