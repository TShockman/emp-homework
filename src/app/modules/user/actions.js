export const NAME = 'user';

export const USER_REQUESTED = `${NAME}/USER_REQUESTED`;
export const USER_FULFILLED = `${NAME}/USER_FULFILLED`;
