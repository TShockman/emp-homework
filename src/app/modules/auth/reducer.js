import {Record} from 'immutable';
import {LOGIN_FULFILLED} from './actions';

// auth state is an immutable record holding only a user id
export const AuthState = Record({
  id: null
});

const initialState = new AuthState();

// on user login, update the state to hold the confirmed id
export default (state = initialState, action) => {
  switch (action.type) {
    case LOGIN_FULFILLED:
      return state.set('id', action.id);
    default: {
      return state;
    }
  }
}