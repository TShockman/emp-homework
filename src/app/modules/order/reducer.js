import {Map} from 'immutable';
import {GET_ORDERS_FULFILLED} from './actions';

// order state is a map from user id to list of order
const initialState = new Map();

export default (state = initialState, action) => {
  switch (action.type) {
    case GET_ORDERS_FULFILLED:
      return state.merge({[action.userId]: action.orders});
    default: {
      return state;
    }
  }
}