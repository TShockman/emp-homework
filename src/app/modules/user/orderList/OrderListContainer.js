import {connect} from 'react-redux';
import OrderList from './OrderList';
import {selectUserState} from '../selectors';
import {USER_ORDERS_REQUESTED} from '../actions';
import {selectAuthState} from '../../auth/selectors';
import {push} from 'redux-little-router';

function mapStateToProps(state) {
  const {orders} = selectUserState(state);
  const {id} = selectAuthState(state)
  return {id, orders};
}

function mapDispatchToProps(dispatch) {
  return {
    requestUserOrders: ({id}) => dispatch({
      type: USER_ORDERS_REQUESTED,
      id
    }),
    push: href => dispatch(push(href))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(OrderList);