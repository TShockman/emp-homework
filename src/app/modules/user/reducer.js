import {Record} from 'immutable';
import {USER_FULFILLED} from './actions';

export const UserState = Record({
  user: null
});

const initialState = new UserState();

export default (state = initialState, action) => {
  switch (action.type) {
    case USER_FULFILLED:
      return state.set('user', action.user);
    default: {
      return state;
    }
  }
}