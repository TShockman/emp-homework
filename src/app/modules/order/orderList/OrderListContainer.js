import {connect} from 'react-redux';
import OrderList from './OrderList';
import {selectUserState} from '../../user/selectors';
import {GET_ORDERS_REQUESTED} from '../actions';
import {selectAuthState} from '../../auth/selectors';
import {push} from 'redux-little-router';
import {selectUserOrders} from '../selectors';

function mapStateToProps(state) {
  const {id} = selectAuthState(state);
  const orders = selectUserOrders(state, id);
  return {id, orders};
}

function mapDispatchToProps(dispatch) {
  return {
    requestUserOrders: ({userId}) => dispatch({
      type: GET_ORDERS_REQUESTED,
      userId
    }),
    push: href => dispatch(push(href))
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(OrderList);