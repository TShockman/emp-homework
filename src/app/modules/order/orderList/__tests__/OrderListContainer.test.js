/* eslint-disable import/first */
jest.mock('../OrderList', () => 'OrderList');

import React from 'react';
import renderer from 'react-test-renderer';
import {createStore} from 'redux';
import {Provider} from 'react-redux';
import {AuthState} from '../../../auth/reducer';
import {NAME as AUTH_NAME} from '../../../auth/actions';
import orderReducer from '../../reducer';
import {NAME} from '../../actions';
import {Map} from 'immutable';

import OrderListContainer from '../index';

describe('OrderListContainer', () => {
  it('Consistently provides OrderList with correct props', () => {
    const store = createStore(orderReducer, {
      [AUTH_NAME]: new AuthState({id: 123}),
      [NAME]: Map({
        '7': [{bogus: 'data A'}],
        '123': [{bogus: 'mock data B'}]
      })
    });
    const tree = renderer.create(
        <Provider store={store}>
          <OrderListContainer/>
        </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
