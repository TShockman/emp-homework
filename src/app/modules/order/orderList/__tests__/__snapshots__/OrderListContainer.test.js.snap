// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderListContainer Consistently provides OrderList with correct props 1`] = `
<OrderList
  id={123}
  orders={
    Array [
      Object {
        "bogus": "mock data B",
      },
    ]
  }
  push={[Function]}
  requestUserOrders={[Function]}
/>
`;
